modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+      next unless ace_body[:flags] && ace_body[:flags][:ace_object_type_present] == 1
+
+      object_type = ace_body[:object_type]
+
+      if ace_access_mask.match(/CONTROL_ACCESS/) && (object_type == CERTIFICATE_ENROLLMENT_EXTENDED_RIGHT || object_type == CERTIFICATE_AUTOENROLLMENT_EXTENDED_RIGHT)
+        if ace_string.match(/DENIED/)
+          flag_allowed_to_enroll = false
+        elsif ace_string.match(/ALLOWED/)
+          flag_allowed_to_enroll = true
+          allowed_sids << ace_body[:sid].to_s
+        end
+      end
+    end
+
+    [flag_allowed_to_enroll, allowed_sids]
+  end
+
+  def query_ldap_server(raw_filter, attributes, base_prefix: nil)
+    ldap_connect do |ldap|
+      validate_bind_success!(ldap)
+
+      if !@base_dn.blank?
+        vprint_status("Using already discovered base DN: #{@base_dn}")
+      elsif (@base_dn = datastore['BASE_DN'])
+        print_status("User-specified base DN: #{@base_dn}")
+      else
+        print_status('Discovering base DN automatically')
+
+        unless (@base_dn = discover_base_dn(ldap))
+          print_warning("Couldn't discover base DN!")
+        end
+      end
+
+      if @base_dn.blank?
+        fail_with(Failure::BadConfig, 'No base DN was found or specified, cannot continue!')
+      end
+
+      if base_prefix.blank?
+        full_base_dn = @base_dn.to_s
+      else
+        full_base_dn = "#{base_prefix},#{@base_dn}"
+      end
+      begin
+        filter = Net::LDAP::Filter.construct(raw_filter)
+      rescue StandardError => e
+        fail_with(Failure::BadConfig, "Could not compile the filter to find the ESC1 objects! Error was #{e}")
+      end
+
+      returned_entries = ldap.search(base: full_base_dn, filter: filter, attributes: attributes)
+      query_result = ldap.as_json['result']['ldap_result']
+
+      result_code, result_message = check_query_result_code(query_result, filter)
+      case result_code
+      when -1
+        fail_with(Failure::BadConfig, result_message)
+      when 0
+        vprint_good(result_message)
+      when 1
+        fail_with(Failure::NoAccess, result_message)
+      when 2
+        fail_with(Failure::UnexpectedReply, result_message)
+      end
+
+      if returned_entries.blank?
+        print_error("No results found for #{filter}.")
+
+        nil
+      else
+
+        returned_entries
+      end
+    end
+  rescue Rex::ConnectionTimeout
+    fail_with(Failure::Unreachable, "Couldn't reach #{datastore['RHOST']}!")
+  rescue Net::LDAP::Error => e
+    fail_with(Failure::UnexpectedReply, "Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+  end
+
+  def convert_sids_to_human_readable_name(sids_array)
+    output = []
+    for sid in sids_array
+      raw_filter = "(objectSID=#{sid})"
+      attributes = ['dn', 'description']
+      sid_entry = query_ldap_server(raw_filter, attributes)
+      fail_with(Failure::UnexpectedReply, 'Could not find any matching SID!') if sid_entry.blank?
```

Comment:
Sounds good, I'll rework this code to take your comments into account and just print out the SID if we can't find any references. I think in most cases LDAP objects aren't deleted but instead are tombstoned so that the references can still exist but are considered inactive, but your point is valid and something I should take into account.

Reply:
Reworked this to just use the SID if no data could be retrieved in https://github.com/rapid7/metasploit-framework/pull/17122/commits/7a2ef23755565bf65de53f21440c00cb4c30325f. Let me know if this looks suitable.